kv/mock.go

```diff
@@ -37,3 +39,78 @@ func (m mockClient) WatchKey(_ context.Context, _ string, _ func(interface{}) bo
 
 func (m mockClient) WatchPrefix(_ context.Context, _ string, _ func(string, interface{}) bool) {
 }
+
+// MockCountingClient is a wrapper around the Client interface that counts the number of times its functions are called.
+// This is used for testing only.
+type MockCountingClient struct {
+	client Client
+
+	mtx   sync.Mutex
+	calls map[string]*atomic.Int32
+}
+
+func NewMockCountingClient(client Client) *MockCountingClient {
+	return &MockCountingClient{
+		client: client,
+		calls:  make(map[string]*atomic.Int32),
+	}
+}
+
+func (mc *MockCountingClient) List(ctx context.Context, prefix string) ([]string, error) {
+	mc.incCall("List")
+
+	return mc.client.List(ctx, prefix)
+}
+func (mc *MockCountingClient) Get(ctx context.Context, key string) (interface{}, error) {
+	mc.incCall("Get")
+
+	return mc.client.Get(ctx, key)
+}
+
+func (mc *MockCountingClient) Delete(ctx context.Context, key string) error {
+	mc.incCall("Delete")
+
+	return mc.client.Delete(ctx, key)
+}
+
+func (mc *MockCountingClient) CAS(ctx context.Context, key string, f func(in interface{}) (out interface{}, retry bool, err error)) error {
+	mc.incCall("CAS")
+
+	return mc.client.CAS(ctx, key, f)
+}
+
+func (mc *MockCountingClient) WatchKey(ctx context.Context, key string, f func(interface{}) bool) {
+	mc.incCall("WatchKey")
+
+	mc.client.WatchKey(ctx, key, f)
+}
+
+func (mc *MockCountingClient) WatchPrefix(ctx context.Context, key string, f func(string, interface{}) bool) {
+	mc.incCall("WatchPrefix")
+
+	mc.client.WatchPrefix(ctx, key, f)
+}
+
+func (mc *MockCountingClient) GetCalls(name string) int {
```

Comment:
I would just use a separate counter member for each type of call instead of a map of them since we know all the types of calls that will be made. There's no need to add new calls dynamically.

e.g.

```go

type MockCountingClient struct {
	client Client

	GetCalls *atomic.Int32
	ListCalls *atomic.Int32
	// ...
}

Reply:
Yeah makes sense actually 😅 . Its much simpler that way. I will update it

Reply:
Addressed here [77843bd](https://github.com/grafana/dskit/pull/618/commits/77843bddc2e67e27cf1f24df14727a2ab29d9789)